doc/memos/rdm-draft-buschsieweke-timer-api.md

```diff
@@ -0,0 +1,201 @@
+| Context of Timer Callback | Thread    | ?         | ?         | Both? (3) | ?         | n.a.      |
+| System Time               | ✓         | ✓         | ✓         | ✓         | ✓         | ✓         |
+| Delays                    | ✓         | ● (1)     | ✓         | ● (4)     | ✓ (5)     | ✓         |
+
+✓ = 1st level API support, ● = supported by combining APIs, ✗ = no support
+
+1. Threads can wait until a software timer fires. Software timers accept NULL
+   as empty callback. Combined, this can be used to delay execution.
+2. Contiki provides with `ctimer_reset()`/`etimer_reset()` which restarts the
+   timer from the previous expiration time, allowing to easily achieve
+   periodic callbacks.
+3. The Rtimer library seems to run the callbacks in IRQ context and the Ctimer
+   library in thread context.
+4. The Etimer library posts events on timer expiration and a thread can block
+   until a specific event has been posted. Combined, this can be used to
+   delay execution
+5. Only available when using TOSTThreads
+
+# 2. Requirements
+
+The following lists of requirements consist of qualitative and quantitative
+requirements. For the quantitative requirements, any improvement will enable
+more scenarios to use the timer API. Thus, no meaningful lower bounds can be
+specified. Instead, a set of benchmarks will be developed that allow to
+continuously improve the performance and to detect any possible regressions of
+future changes.
+
+## Feature Requirements
+
+- Software Timers
+    - Upper bound for number of timers only given by available resources
+    - APIs to set and abort timers
+    - Support for periodic timers (support via explicit calls to some
+      `timer_reset()` like in Contiki is fine)
+    - Long term low resolution software timers that play well with power management
+    - Short term high resolution software timers
+- Delay of program execution
+    - API to block calling thread for long periods and allows other threads to run
+    - Long term delays must play well with power management
+    - Highly precise short term delays (e.g. for bit banging)
+- System time
+    - APIs to get the current system time in reasonable units
+- Utilities for convenience
+    - Timer utilities
+        - Timer that send a message upon expiration
+        - Timer that post an event upon expiration (to allow execution in thread context)
+        - Timer that wakes up a thread upon expiration
+    - Timeout utilities
+        - Timeout for receiving a message
+        - Timeout for locking a mutex
+
+## Real-Time Requirements
+
+- The implementation must only disable IRQs for short periods of time
+- The implementation must cause as little IRQs as possible and its ISRs must
+  be as short as possible
+- The implementation should provide means to execute callbacks in thread
+  context
+
+## Precision Requirements
+
+- Timer callbacks should never fire prematurely (+- 1 tick of the underlying
```

Comment:
Important point. Is "-1" acceptable?

Reply:
With that formulation, clearly yes. I think the effort needed to prevent this could be quite significant. So if no one cares about it triggering early by 1 tick enough to speak up, we should just allow this :-)

Reply:
> I think the effort needed to prevent this could be quite significant.

`set(val) { val += 1; ... }`, or `run_callback() { while(now() < target_time) {}; ...` can achieve this.

We had a discussion some time ago, where people convinced me that "at least as" semantics are the only sane ones. Don't care about one tick? Set the timeout one less. (Having written that, I realize it also works the other way. :) )

The impact of this is very different depending on us, ms, or even second ticks are used.

I'm OK with keeping `-1` for now.